components/percona_telemetry/worker.cc

```diff
@@ -0,0 +1,91 @@
+#include <mutex>
+
+#include "logger.h"
+#include "worker.h"
+
+Worker::Worker(Config &config, Storage &storage, DataProvider &data_provider,
+               Logger &logger)
+    : config_(config),
+      storage_(storage),
+      data_provider_(data_provider),
+      logger_(logger),
+      stop_worker_thd_(false),
+      caller_active_(ATOMIC_FLAG_INIT) {}
+
+bool Worker::start() {
+  std::thread thd(&Worker::worker_thd_fn, this);
+  std::swap(thd_, thd);
+  return false;
+}
+
+bool Worker::stop() {
+  if (caller_active_.test_and_set()) {
+    logger_.info("worker active. unload prohibited");
+    return true;
+  }
+  stop_worker_thd_.store(true);
+  cv_.notify_all();
+  thd_.join();
+
+  return false;
+}
+
+void Worker::worker_thd_fn() {
+  std::mutex m;
+
+  int grace_interval = config_.grace_interval();
+  logger_.info("Applying Telemetry grace interval %d seconds", grace_interval);
+  {
+    std::unique_lock<std::mutex> lock(m);
+    if (std::cv_status::timeout !=
+        cv_.wait_for(lock, std::chrono::seconds(grace_interval))) {
+      // the thread was signaled to exit
+      return;
+    }
+  }
+
+  data_provider_.thread_access_begin();
+
+  // TODO:
+  // The server is probably still initializing if the grace period is low, so
```

Comment:
Yes, I need to think more about it. In fact this will be executed after the grace period which is 24 hours, so everything should be initialized, but of course we need to handle this corner case.

Reply:
Implemented graceful handling of this situation.